### PR TITLE
[tests] Make rpcauth.py testable and add unit tests

### PR DESCRIPTION
**Pull request by nixbox**
refs #12995

### CODE REVIEW

**Comment by laanwj** [comment]
Some lint issues:
```
./test/util/rpcauth-test.py:19:1: E402 module level import not at top of file
./test/util/rpcauth-test.py:20:1: E402 module level import not at top of file
./test/util/rpcauth-test.py:21:1: E402 module level import not at top of file
./test/util/rpcauth-test.py:22:1: E402 module level import not at top of file
^---- failure generated from contrib/devtools/lint-python.sh
```
Also the test fails to launch:
```
/usr/bin/python3.4 ../test/util/rpcauth-test.py
/usr/bin/python3.4: can't open file '../test/util/rpcauth-test.py': [Errno 2] No such file or directory
```

**Comment by nixbox** [comment]
Fixed the lint issues.

The reason you are not able to run the test manually is probably because config.ini is missing under test/

After running ./configure, I am able to run the test manually as well.

**Comment by laanwj** [comment]
utACK ebff14786b7addd4bcc6e74691394a77019f2bf6

File: Makefile.am [hunk 1]
```diff
@@ -225,6 +225,7 @@ EXTRA_DIST += \
     test/util/data/blanktxv1.json \
     test/util/data/blanktxv2.hex \
     test/util/data/blanktxv2.json \
+    test/util/rpcauth-test.py \
```
Comment (jnewbery) [inline Makefile.am]: nit: seems like somewhat random placement. Why not alphabetically sorted like the rest?

File: share/rpcauth/rpcauth.py [hunk 2]
```diff
@@ -9,26 +9,36 @@
...
-if len(sys.argv) < 2:
-    sys.stderr.write('Please include username as an argument.\n')
-    sys.exit(0)
+def generate_salt():
+    # This uses os.urandom() underneath
+    cryptogen = SystemRandom()
 
-username = sys.argv[1]
+    # Create 16 byte hex salt
+    salt_sequence = [cryptogen.randrange(256) for i in range(16)]
```
Comment (jnewbery) [inline share/rpcauth/rpcauth.py]: nit: prefer `_` for unused loop variables.

File: share/rpcauth/rpcauth.py [hunk 3]
```diff
@@ -9,26 +9,36 @@
...
+    return ''.join([format(r, 'x') for r in salt_sequence])
 
-#This uses os.urandom() underneath
-cryptogen = SystemRandom()
+def generate_password(salt):
+    # Create 32 byte b64 password
```
Comment (jnewbery) [inline share/rpcauth/rpcauth.py]: nit: prefer docstrings for function-level comments.

File: share/rpcauth/rpcauth.py [hunk 4]
```diff
@@ -9,26 +9,36 @@
...
+        sys.stderr.write('Please include username as an argument.\n')
+        sys.exit(0)
 
-print("String to be appended to bitcoin.conf:")
-print("rpcauth="+username+":"+salt+"$"+result)
-print("Your password:\n"+password)
+    username = sys.argv[1]
+
+    salt = generate_salt()
+    (password, password_hmac) = generate_password(salt)
```
Comment (jnewbery) [inline share/rpcauth/rpcauth.py]: No need to use parens here. Just:

```
password, password_hmac = generate_password(salt)
```

(you could also change the return statement in `generate_password()` to `return password, password_hmac`, but under the covers Python will construct a tuple to return anyway, so it's equivalent.)

File: share/rpcauth/rpcauth.py [hunk 5]
```diff
@@ -9,26 +9,36 @@
...
+
+    print('String to be appended to bitcoin.conf:')
+    print('rpcauth='+username+':'+salt+'$'+password_hmac)
```
Comment (jnewbery) [inline share/rpcauth/rpcauth.py]: nit: prefer string formatters (`{}` or `%` - we can't use f-strings yet).

If you must concatenate strings, please surround the `+` with spaces so the code isn't so difficult to parse.

File: test/util/rpcauth-test.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
+# Copyright (c) 2015-2018 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test share/rpcauth/rpcauth.py
+"""
+import base64
+try:
+    import configparser
+except ImportError:
```
Comment (jnewbery) [inline test/util/rpcauth-test.py]: I think this is unnecessary - this is for python2 compatibility. We now no longer support python2 across the project.

File: test/util/rpcauth-test.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import importlib
+import os
+import sys
+import unittest
+
+class TestRPCAuth(unittest.TestCase):
+    def setUp(self):
+        config = configparser.ConfigParser()
+        file_path = os.path.abspath(os.path.dirname(__file__)) + "/../config.ini"
+        config_file = open(file_path)
```
Comment (jnewbery) [inline test/util/rpcauth-test.py]: nit: prefer `with open() as:` for opening files.

**Review by jnewbery — COMMENTED** [review]
A few nits inline after testing.

File: Makefile.am (hunk 1 above)
Comment (nixbox) [reply Makefile.am]: Done

File: share/rpcauth/rpcauth.py (hunk 2 above)
Comment (nixbox) [reply share/rpcauth/rpcauth.py]: Done

File: share/rpcauth/rpcauth.py (hunk 3 above)
Comment (nixbox) [reply share/rpcauth/rpcauth.py]: Done

File: share/rpcauth/rpcauth.py (hunk 4 above)
Comment (nixbox) [reply share/rpcauth/rpcauth.py]: Done

File: share/rpcauth/rpcauth.py (hunk 5 above)
Comment (nixbox) [reply share/rpcauth/rpcauth.py]: Done

File: test/util/rpcauth-test.py (hunk 6 above)
Comment (nixbox) [reply test/util/rpcauth-test.py]: Done

File: share/rpcauth/rpcauth.py [hunk 8]
```diff
@@ -9,26 +9,36 @@
...
+    cryptogen = SystemRandom()
 
-username = sys.argv[1]
+    # Create 16 byte hex salt
+    salt_sequence = [cryptogen.randrange(256) for _ in range(16)]
+    return ''.join([format(r, 'x') for r in salt_sequence])
 
-#This uses os.urandom() underneath
-cryptogen = SystemRandom()
+"""Create 32 byte b64 password"""
```
Comment (jnewbery) [inline share/rpcauth/rpcauth.py]: Docstrings should be the first statement *inside* the function (https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring)

File: test/util/rpcauth-test.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import hmac
+import importlib
+import os
+import sys
+import unittest
+
+class TestRPCAuth(unittest.TestCase):
+    def setUp(self):
+        config = configparser.ConfigParser()
+        file_path = os.path.abspath(os.path.dirname(__file__)) + "/../config.ini"
```
Comment (jnewbery) [inline test/util/rpcauth-test.py]: nit: use `os.path.join()` for maximum portability (instead of concat'ing path strings)

**Review by jnewbery — COMMENTED** [review]
couple more nits inline

File: test/util/rpcauth-test.py (hunk 9 above)
Comment (MarcoFalke) [reply test/util/rpcauth-test.py]: Also, should probably say `config_path`

File: test/config.ini.in [hunk 10]
```diff
@@ -9,6 +9,7 @@
 SRCDIR=@abs_top_srcdir@
 BUILDDIR=@abs_top_builddir@
 EXEEXT=@EXEEXT@
+RPCAUTH=@abs_top_srcdir@/share/rpcauth/rpcauth.py
```
Comment (MarcoFalke) [inline test/config.ini.in]: Could do a rebase, since this is already in master?

File: share/rpcauth/rpcauth.py (hunk 8 above)
Comment (nixbox) [reply share/rpcauth/rpcauth.py]: Done

File: test/util/rpcauth-test.py (hunk 9 above)
Comment (nixbox) [reply test/util/rpcauth-test.py]: Done

File: test/config.ini.in (hunk 10 above)
Comment (nixbox) [reply test/config.ini.in]: Done

**Comment by jnewbery** [comment]
Tested ACK 6674a75bfbdf680d0a3648f44c6591ff5fc983cf.

Travis failure was unrelated failure in `mining_prioritisetransaction.py`. I've restarted the job.

**Comment by nixbox** [comment]
All green, ready to go?

**Pull request merged**